### PR TITLE
feat: login/logout feature

### PR DESCRIPTION
**Pull request by qoosuperman**
使用 warden 來實作 login / logout feature

參考的文章：
1. https://rst.software/blog/2017/10/simple-warden-authentication-service/
2. https://medium.com/@altanner/react-user-authentication-with-rails-sessions-and-redux-194b5d31fe5a

### CODE REVIEW

File: config/application.rb [hunk 1]
```diff
@@ -26,5 +26,9 @@ class Application < Rails::Application
     # Skip views, helpers and assets when generating a new resource.
     config.api_only = true
     config.i18n.default_locale = 'zh-TW'
+
+    config.session_store :cookie_store, key: '_taverneer_sesson', expire_after: 1.day
+    config.middleware.use ActionDispatch::Cookies
+    config.middleware.use config.session_store, config.session_options
```
Comment (qoosuperman) [inline config/application.rb:34]: 參考 [這篇文章](https://medium.com/@altanner/react-user-authentication-with-rails-sessions-and-redux-194b5d31fe5a) 的 session store 設定

File: app/controllers/api/v1/sessions_controller.rb [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+        sign_in(user)
+        respond_with_json(status: :created, message: 'success!')
+      end
+
+      def destroy; end
+
+      private
+
+      def sign_in(user)
+        # Do nothing. User already signed in and we are not forcing it.
```
Comment (choznerol) [inline app/controllers/api/v1/sessions_controller.rb:28]: > ... and we are not forcing it.

不太懂這句 🙇

File: spec/requests/api/v1/sessions_controller_spec.rb [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    subject { post api_v1_users_sign_in_path, params: params }
+
+    let!(:user) do
+      create(:user, email: 'coolguy@gmail.com', password: 'password1234')
+    end
+
+    context 'when params is valid' do
+      let(:params) do
+        { user: {
+          name: 'test user',
```
Comment (choznerol) [inline spec/requests/api/v1/sessions_controller_spec.rb:14]:
```suggestion
```
session#create 的實作好像其實不會看 `name`?

File: spec/requests/api/v1/sessions_controller_spec.rb [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+          email: 'coolguy@gmail.com',
+          password: 'password1234'
+        } }
+      end
+
+      it 'gets created response' do
+        subject
+        expect(response).to have_http_status(:created)
+        expect(json_body).to eq({
+                                  status: 'created',
```
Comment (choznerol) [inline spec/requests/api/v1/sessions_controller_spec.rb:24]: NIT: failure app 那邊的 body 的 statue attribute 是 status code，這邊則是 status text，會想統一嗎?

File: app/controllers/concerns/authentication/helper_methods.rb [hunk 5]
```diff
@@ -0,0 +1,15 @@
+# frozen_string_literal: true
+
+module Authentication
+  module HelperMethods
+    delegate :authenticate!, to: :warden
+
```
Comment (choznerol) [inline app/controllers/concerns/authentication/helper_methods.rb:6]:
```suggestion
```
sessions#create 裡實際上仍是寫 `warden.authenticate!`，這邊其實沒用到?

File: app/controllers/api/v1/sessions_controller.rb (hunk 2 above)
Comment (qoosuperman) [reply app/controllers/api/v1/sessions_controller.rb:28]: 我改成中文的好了ＸＤ
是從 [這裡](https://github.com/heartcombo/devise/blob/main/lib/devise/controllers/sign_in_out.rb#L50) 抄來的

File: config/initializers/session_store.rb [hunk 6]
```diff
@@ -0,0 +1,9 @@
+# frozen_string_literal: true
+
+Rails.application.config.session_store :cookie_store, {
+  key: '_taverneer',
+  domain: :all,
+  same_site: :none,
```
[the remark this answers is not in the document]
Comment (qoosuperman) [reply config/initializers/session_store.rb]: fixed [here](https://github.com/qoosuperman/taverneer-backend/pull/6/commits/a92685c519dc33d5c2b989be9b1e6da04c6e68d1#diff-bb2a4b5fcffb49c9f95d0f839ba62c674a05fe17ac90798a38178a8a3a3b528bR3-R7)

File: app/controllers/api/v1/sessions_controller.rb (hunk 2 above)
Comment (qoosuperman) [reply app/controllers/api/v1/sessions_controller.rb]: [fixed](https://github.com/qoosuperman/taverneer-backend/pull/6/commits/2aedb50b907e6e386ae8ca2f4e6f7e3da5a29bdd#diff-bbe4b7d878668139956c768736e4a042fda018c947147c2b84f12db0ca3c73e9R17-R18)

File: app/controllers/concerns/authentication/helper_methods.rb (hunk 5 above)
Comment (qoosuperman) [reply app/controllers/concerns/authentication/helper_methods.rb]: [removed](https://github.com/qoosuperman/taverneer-backend/pull/6/commits/2aedb50b907e6e386ae8ca2f4e6f7e3da5a29bdd#diff-a2a89d8ca5e17d1c49db867ffb80c6558ff1cff80aa5444782f33c50348ebfbcR3-R7)

File: spec/requests/api/v1/sessions_controller_spec.rb (hunk 3 above)
Comment (qoosuperman) [reply spec/requests/api/v1/sessions_controller_spec.rb]: [removed](https://github.com/qoosuperman/taverneer-backend/pull/6/commits/168d19326fb32a8d8c2f38e95de7d8e355915057#diff-a09ec63662d36cff296195e53a1d957e23c1de79454dada010fbaa01b47ccce9R11-R17)

File: app/controllers/api/v1/sessions_controller.rb (hunk 2 above)
Comment (choznerol) [reply app/controllers/api/v1/sessions_controller.rb]: > 是從 [這裡](https://github.com/heartcombo/devise/blob/main/lib/devise/controllers/sign_in_out.rb#L50) 抄來的

原來如此，那可能是指原 code 的 `options.delete(:force)`

**Review by choznerol — COMMENTED** [review]
LGTM

**Pull request merged**